app_flutter/lib/status_grid.dart

```diff
@@ -45,29 +44,128 @@ class StatusGridContainer extends StatelessWidget {
+
+    /// O(Tasks * CommitStatuses). In production, this is usually O(85 * 100) ~ 8500 operations.
+    for (CommitStatus status in statuses) {
+      for (Stage stage in status.stages) {
+        for (Task task in stage.tasks) {
+          final String key = _taskColumnKey(task);
+          if (taskColumnKeyIndex.containsKey(key)) {
+            continue;
+          }
+
+          taskColumnKeyIndex[key] = currentIndex++;
+        }
+      }
+    }
+
+    return taskColumnKeyIndex;
+  }
+
+  /// Create a matrix of [Task] for easier sorting of [List<CommitStatus>].
+  @visibleForTesting
+  static List<List<Task>> createTaskMatrix(
+      List<CommitStatus> statuses, Map<String, int> taskColumnKeyIndex) {
+    /// Rows are commits, columns are [Task] with same [taskColumnKey].
+    final List<List<Task>> taskMatrix = List<List<Task>>.generate(
+        statuses.length, (_) => List<Task>(taskColumnKeyIndex.keys.length));
+
+    for (int statusIndex = 0; statusIndex < statuses.length; statusIndex++) {
+      final CommitStatus status = statuses[statusIndex];
+      final List<Task> statusTasks = taskMatrix[statusIndex];
+
+      /// Organize [Task] in [CommitStatus] to the column they map to.
+      for (Stage stage in status.stages) {
+        for (Task task in stage.tasks) {
+          statusTasks[taskColumnKeyIndex[_taskColumnKey(task)]] = task;
+        }
+      }
+    }
+
+    return taskMatrix;
+  }
+
+  /// Create [List<Task>] for [List<TaskIcon>] at the top of [StatusGrid].
+  @visibleForTesting
+  static List<Task> createTaskIconRow(
+      List<CommitStatus> statuses, Map<String, int> taskColumnKeyIndex) {
+    final List<Task> taskIconRow = List<Task>(taskColumnKeyIndex.keys.length);
+
+    for (int statusIndex = 0; statusIndex < statuses.length; statusIndex++) {
+      final CommitStatus status = statuses[statusIndex];
+      for (Stage stage in status.stages) {
+        for (Task task in stage.tasks) {
+          final int taskIndex = taskColumnKeyIndex[_taskColumnKey(task)];
+          if (taskIconRow[taskIndex] == null) {
+            taskIconRow[taskIndex] = task;
+          }
+        }
+      }
+    }
+
+    return taskIconRow;
+  }
 }
 
 /// Display results from flutter/flutter repository's continuous integration.
 ///
 /// Results are displayed in a matrix format. Rows are commits and columns
 /// are the results from tasks.
 class StatusGrid extends StatelessWidget {
-  const StatusGrid({Key key, @required this.statuses}) : super(key: key);
+  const StatusGrid(
+      {Key key,
+      @required this.statuses,
+      @required this.taskMatrix,
+      @required this.taskIconRow})
+      : super(key: key);
 
   /// The build status data to display in the grid.
   final List<CommitStatus> statuses;
 
+  /// Computed 2D array of [Task] to make it easy to retrieve and sort tasks.
+  final List<List<Task>> taskMatrix;
+
+  /// [Task] that can be used for the row of [TaskIcon].
+  final List<Task> taskIconRow;
```

Comment:
Nit, there is more than one  task here: `A list of [Task] objects that ...`

Reply:
Good catch! Fixed